flink-end-to-end-tests/flink-dataset-fine-grained-recovery-test/src/main/java/org/apache/flink/batch/tests/util/FileBasedOneShotLatch.java

```diff
@@ -0,0 +1,126 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.batch.tests.util;
+
+import com.sun.nio.file.SensitivityWatchEventModifier;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A synchronization aid that allows a single thread to wait on the creation of a specified file.
+ */
+@NotThreadSafe
+public class FileBasedOneShotLatch implements Closeable {
+
+	private final Path latchFile;
+
+	private final WatchService watchService;
+
+	private boolean released;
+
+	public FileBasedOneShotLatch(final Path latchFile) {
+		this.latchFile = checkNotNull(latchFile);
+
+		final Path parentDir = checkNotNull(latchFile.getParent(), "latchFile must have a parent");
+		this.watchService = initWatchService(parentDir);
+	}
+
+	private static WatchService initWatchService(final Path parentDir) {
+		final WatchService watchService = createWatchService();
+		watchForLatchFile(watchService, parentDir);
+		return watchService;
+	}
+
+	private static WatchService createWatchService() {
+		try {
+			return FileSystems.getDefault().newWatchService();
+		} catch (IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	private static void watchForLatchFile(final WatchService watchService, final Path parentDir) {
+		try {
+			parentDir.register(
+				watchService,
+				new WatchEvent.Kind[]{StandardWatchEventKinds.ENTRY_CREATE},
+				SensitivityWatchEventModifier.HIGH);
+		} catch (IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	/**
+	 * Waits until the latch file is created.
+	 *
+	 * @throws InterruptedException if interrupted while waiting
+	 */
+	public void await() throws InterruptedException {
+		if (isReleasedOrReleasable()) {
+			return;
+		}
+
+		awaitLatchFile(watchService);
+	}
+
+	private void awaitLatchFile(final WatchService watchService) throws InterruptedException {
+		while (true) {
+			WatchKey take = watchService.take();
+			if (isReleasedOrReleasable()) {
```

Comment:
It could be that other files are created in the directory.

Reply:
ok. I suppose one _could_ check this via `WatchKey#pollEvents#Context, but the version in the PR is wayyyy simpler. Prone to files being deleted in-between, but this seems unlikely.

Reply:
> Prone to files being deleted in-between, but this seems unlikely.

True, didn't think about that case. I think it's acceptable to leave it as is because:
- we wait until the job is finished before deleting files
- class is contained in this module
- code is simpler (as you already mentioned)